src/compat/lt-127/ArkoudaRandomCompat.chpl

```diff
@@ -0,0 +1,3015 @@
+     these algorithms.
+   */
+  enum RNG {
+    PCG = 1,
+    NPB = 2
+  }
+
+  /* The default RNG. The current default is PCG - see :mod:`PCGRandom`. */
+  param defaultRNG = RNG.PCG;
+
+  type RandomStream = if defaultRNG == RNG.PCG then PCGRandomStream
+                                               else NPBRandomStream;
+
+  // CHPLDOC FEEDBACK: If easy, I'd suggest either deprecating the
+  // :arg <type> <name>: form or else switching the order to
+  // :arg <name> <type>: as it's an easy trap to simply type the same
+  // argument as in the function's signature.
+
+
+  // CHPLDOC FIXME: if the first line below is shifted one character to the
+  // left, it ends up being rendered like a method
+
+  // CHPLDOC FIXME: if the first line below is shifted one character to the
+  // right, it ends up causing a warning (promoted to error) in the .rst
+  // file.  It'd be preferable to have it declare the issue in terms of
+  // the .chpl line numbers.
+  //
+
+  private
+  proc isSupportedNumericType(type t) param
+    return isNumericType(t) || isBoolType(t);
+
+  /*
+
+    Fills a rectangular array of numeric elements with pseudorandom values in parallel using
+    a new stream implementing :class:`RandomStreamInterface` created
+    specifically for this call.  The first `arr.size` values from the stream
+    will be assigned to the array's elements in row-major order. The
+    parallelization strategy is determined by the array.
+
+    .. note::
+      :mod:`NPBRandom` only supports `real(64)`, `imag(64)`, and `complex(128)`
+      numeric types. :mod:`PCGRandom` supports all primitive numeric types.
+
+    :arg arr: The array to be filled, where T is a primitive numeric type. Only
+      rectangular arrays are supported currently.
+    :type arr: `[] T`
+
+    :arg seed: The seed to use for the PRNG.  Defaults to
+     `oddCurrentTime` from :type:`RandomSupport.SeedGenerator`.
+    :type seed: `int(64)`
+
+    :arg algorithm: A param indicating which algorithm to use. Defaults to :param:`defaultRNG`.
+    :type algorithm: :type:`RNG`
+  */
+  proc fillRandom(arr: [], seed: int(64) = SeedGenerator.oddCurrentTime, param
+      algorithm=defaultRNG)
+    where isSupportedNumericType(arr.eltType) {
+    var randNums = createRandomStream(seed=seed,
+                                      eltType=arr.eltType,
+                                      parSafe=false,
+                                      algorithm=algorithm);
+    randNums.fillRandom(arr);
+  }
+
+  proc fillRandom(arr: [], min: arr.eltType, max: arr.eltType, seed: int(64) = SeedGenerator.oddCurrentTime) 
+    where isSupportedNumericType(arr.eltType) {
+      var randNums = createRandomStream(seed=seed,
+                                      eltType=arr.eltType,
+                                      parSafe=false,
+                                      algorithm=defaultRNG);
+      randNums.fillRandom(arr, min, max);
+  }

```

Comment:
@ronawho @bmcdonald3 @bradcray I had to add this function to our version of Random.chpl (Which @bradcray mentioned, so I was expecting to have to) which would allow for the bounded variation of `fillRandom()` to be called. I tried quite a few attempts at creating the function outside of Random.chpl, or calling the bounded version directly from our code, so it wouldn't become a compatibility issue, but as far as I could tell it can only work from the standalone definition created within the module. Is there something I'm missing here which would allow me to call the bounded version of `fillRandom` directly from arkouda's code without having to add the helper function within the compat file? My concern is that by adding this function here, it means we would have to continue using the compatibility module even when using Chapel 1.27 due to Chapel 1.27 not having this function.

Reply:
@joshmarshall1 : Sorry for the slow response here today, and I'm not going to be able to spend much time on it before bed.  But in hopes that it helps:

I hadn't done anything with the branch I mentioned in https://github.com/Bears-R-Us/arkouda/issues/1220#issuecomment-1115257667 since posting https://github.com/Bears-R-Us/arkouda/issues/1220#issuecomment-1148145256, but just returned to it today and found that it was pretty easy to update it now that 1.27.0 has a standalone fillRandom() routine.  I've pushed that update back to GitHub, so am hoping that by comparing what I did with what you've done, you'll be able to sort out why my RandomHelp.chpl only needed to be 228 lines rather than O(|Random.chpl|) size.  I can potentially spend more time on this tomorrow if you have specific questions I can help with, or my comments in the first issue link above on how I ended up with this solution aren't clear (I'd need to spend more time with them myself to remember what I did and why).

With that, I was able to:

* compile with 1.26.0
* almost compile with Chapel `main` (thinking of it as "almost 1.27.0"), until I hit the issue that #1555 is designed to fix.
* compile with 1.27.0

I should emphasize, "compile"—I didn't do testing beyond that due in part to the hour, in part not having a very good sense what I would/should test.

Reply:
I was able to get this working using your RandomHelp.chpl. As far as I can tell, this is still going to be needed for chapel 1.27.0. I wasn't able to find a native way in 1.27.0 to call the bounded version of fillRandom(). Is that accurate?

Reply:
If I'm understanding your question correctly, you're right that the `RandomHelp.chpl` module (or equivalent) would still be required for Chapel 1.27.0, but just for the sake of resolving the module name (and, for me, the fact that it's nearly empty makes that seem like not a big deal—in fact looking at it again today, I think even the `use Random` can probably be removed?).  However, once Arkouda is no longer supporting 1.26.0, it (and its uses/imports in Arkouda) can be removed completely.  I think that this is consistent with what we're doing for other shims that involve changing the behavior of a pre-existing module rather than just adding new modules, though I could be wrong about that.  Tagging @ronawho, who's probably created the most of them.

[That note makes me think: I wonder whether there are now old compatibility modules that could now be removed—for example, it looks to me like the three copies of `ArkoudaRegexCompat.chpl` could _potentially_ be removed at this point?]

The other approach would be to name the compatibility module `Random.chpl` in 1.26.0, cloning the 1.27.0 Random.chpl file into that helper module (assuming it doesn't have other features that would break the 1.26.0 compiler), but... that seems a lot less satisfying to me in terms of code duplication and the potential fragility noted here.  And it will generate the warning telling how it broke the module ambiguity as you noted.  I don't think we have a way to squash that warning currently, though even if we did (say `--no-warn-module-ambiguities`) it feels like a scary flag to throw in terms of potentially missing other surprises, at least without tying it to a specific module (say `--no-warn-module-ambiguities=Random`).

Hope this is helpful.

Reply:
> I think that this is consistent with what we're doing for other shims that involve changing the behavior of a pre-existing module rather than just adding new modules, though I could be wrong about that. Tagging @ronawho, who's probably created the most of them.

I agree

> [That note makes me think: I wonder whether there are now old compatibility modules that could now be removed—for example, it looks to me like the three copies of ArkoudaRegexCompat.chpl could potentially be removed at this point?]

https://github.com/Bears-R-Us/arkouda/issues/1547 was opened to discuss what versions we still want to support and I think once that settles we'll remove support for older compat shims.